expression/scalar_function.go

```diff
@@ -438,13 +439,119 @@ func (sf *ScalarFunction) EvalJSON(ctx sessionctx.Context, row chunk.Row) (types
 
 // HashCode implements Expression interface.
 func (sf *ScalarFunction) HashCode(sc *stmtctx.StatementContext) []byte {
+	if sc.CanonicalHashCode.Load() {
+		simpleCanonicalizedHashCode(sf, sc)
+		return sf.hashcode
+	}
 	if len(sf.hashcode) > 0 {
 		return sf.hashcode
 	}
 	ReHashCode(sf, sc)
 	return sf.hashcode
 }
 
+func ExpressionsSemanticEqual(ctx sessionctx.Context, expr1, expr2 Expression) bool {
+	sc := ctx.GetSessionVars().StmtCtx
+	sc.CanonicalHashCode.Store(true)
+	defer sc.CanonicalHashCode.Store(false)
+	return string(expr1.HashCode(sc)) == string(expr2.HashCode(sc))
```

Comment:
Using `bytes.Equal(...)` seems better. `string(...)` can incur some unnecessary data copy and object allocation.